### PR TITLE
raftstore: check if tick generates any readiness

### PR DESCRIPTION
**Pull request by BusyJay**
So handle_raft_ready will be called only when it's necessary.

@siddontang @hhkbp2 @zhangjinpeng1987 PTAL

### CODE REVIEW

File: src/raft/raft.rs [hunk 1]
```diff
@@ -611,24 +611,29 @@ impl<T: Storage> Raft<T> {
 
     // tick_election is run by followers and candidates after self.election_timeout.
     // TODO: revoke pub when there is a better way to test.
-    pub fn tick_election(&mut self) {
+    pub fn tick_election(&mut self) -> bool {
```
Comment (ngaut) [inline src/raft/raft.rs]: Add comments for return value.

File: src/raft/raft.rs [hunk 2]
```diff
@@ -611,24 +611,29 @@ impl<T: Storage> Raft<T> {
...
+
+        self.election_elapsed = 0;
+        let m = new_message(INVALID_ID, MessageType::MsgHup, Some(self.id));
+        self.step(m).is_ok();
+        true
     }
 
     // tick_heartbeat is run by leaders to send a MsgBeat after self.heartbeat_timeout.
-    fn tick_heartbeat(&mut self) {
+    fn tick_heartbeat(&mut self) -> bool {
```
Comment (ngaut) [inline src/raft/raft.rs]: Ditto.

File: src/raft/raw_node.rs [hunk 3]
```diff
@@ -227,8 +227,8 @@ impl<T: Storage> RawNode<T> {
     }
 
     // Tick advances the internal logical clock by a single tick.
-    pub fn tick(&mut self) {
-        self.raft.tick();
+    pub fn tick(&mut self) -> bool {
```
Comment (ngaut) [inline src/raft/raw_node.rs]: Ditto.

Comment (ngaut) [inline src/raft/raw_node.rs]: I think we should also commit to etcd.

**Comment by siddontang** [comment]
Maybe we should handle this in `on_raft_message` and `propose` too.  A region may be added to pending_raft_group many times in these functions in one mio tick.

**Comment by BusyJay** [comment]
`propose` has been handled properly. `step` function is complicated, it's more convenient to detect changes via has_ready in this case.

**Comment by siddontang** [comment]
> step function is complicated

Sorry, what I mean is here: 

```
        let peer = self.region_peers.get_mut(&region_id).unwrap();
        try!(peer.step(msg.take_message()));

        // Add into pending raft groups for later handling ready.
        self.pending_raft_groups.insert(region_id);
```

We can use `mark_to_be_checked` too.

File: src/raft/raft.rs [hunk 4]
```diff
@@ -614,24 +616,31 @@ impl<T: Storage> Raft<T> {
...
+
+        self.election_elapsed = 0;
+        let m = new_message(INVALID_ID, MessageType::MsgHup, Some(self.id));
+        self.step(m).is_ok();
+        true
     }
 
     // tick_heartbeat is run by leaders to send a MsgBeat after self.heartbeat_timeout.
-    fn tick_heartbeat(&mut self) {
+    // Return true to indicate that handle_raft_ready needs to be called.
```
Comment (hhkbp2) [inline src/raft/raft.rs]: I'm not sure whether it's a good idea to write comment like this mentioning the interface of a outer module which calls `raw_node.tick()`.

Comment (siddontang) [inline src/raft/raft.rs]: yes, not proper here. maybe can write `indicate that processing a raft step?` or others.

File: src/raft/raft.rs [hunk 5]
```diff
@@ -603,7 +603,9 @@ impl<T: Storage> Raft<T> {
         self.maybe_commit();
     }
 
-    pub fn tick(&mut self) {
+    /// Return true to indicate that some internal states have been changed that
```
Comment (hhkbp2) [inline src/raft/raft.rs]: Strictly speaking, it's possible that the state doesn't change after `tick_heartbeat() -> step MsgCheckQuorum()`.

**Comment by hhkbp2** [comment]
Rest LGTM

**Pull request merged**